cmd/gcp-controller-manager/csr_signer.go

```diff
@@ -44,17 +43,23 @@ var (
 // ClusterSigningGKERetryBackoff is the backoff between GKE cluster signing retries.
 const ClusterSigningGKERetryBackoff = 500 * time.Millisecond
 
+type csrClient interface {
+	UpdateStatus(context.Context, *capi.CertificateSigningRequest, metav1.UpdateOptions) (*capi.CertificateSigningRequest, error)
+}
+
 // gkeSigner uses external calls to GKE in order to sign certificate signing
 // requests.
 type gkeSigner struct {
 	webhook      *webhook.GenericWebhook
 	ctx          *controllerContext
 	retryBackoff time.Duration
 	validators   []csrValidator
+
+	csrs csrClient
```

Comment:
why replicate the client in csrs and in ctx?

Reply:
I wanted to take this functionality in terms of a local interface to make testing easier.  My thinking was that I would follow this PR up with some work to break apart the big controllerContext object; I think it's hurting more than helping.

If I write all the functionality here in terms of a "top-level" kubernetes client, how do I go about testing it without defining a bunch of local mocks, one for each level of the client.This().That().AndTheOther() chain?

Reply:
Ah, now I see fake.Clientset.  Let me give that a try

Reply:
if you just want to echo the updated object on update:
```
			fakeClient := fake.NewSimpleClientset()
			fakeClient.PrependReactor("update", "certificatesigningrequests", func(action ktesting.Action) (bool, runtime.Object, error) {
				if update, ok := action.(ktesting.UpdateAction); ok {
					return true, update.GetObject(), nil
				}
				return false, nil, nil
			})
			ctlCtx := &controllerContext{
				client:                      fakeClient,
				clusterSigningGKEKubeconfig: kubeConfig.Name(),
				recorder:                    record.NewFakeRecorder(10),
			}
```

Reply:
OK, I have switched back to just taking a client.Interface, and using fake.NewSimpleClientset to mock out the behavior.

It appears to work properly without adding a custom reactor; if I add code that checks the returned CSR object, it does have the certificate attached, as expected.